src/api/PluginManager.ts

```diff
@@ -113,18 +59,77 @@ export function startCorePlugins() {
     for (const pluginClass of [Badges, CommandHandler, CoreCommands, NoTrack]) {
         const { name } = pluginClass;
         try {
-            logger.info("Loading CorePlugin: " + name);
-            new pluginClass({ name, description: "", version: "1.0.0", authors: [{ name: "Aliucord", id: "000000000000000000" }] }).start();
+            logger.info(`Starting core plugin ${name}`);
+            new pluginClass({
+                name,
+                description: "",
+                version: "1.0.0",
+                authors: [{ name: "Aliucord", id: "000000000000000000" }]
+            }).start();
         } catch (e) {
-            logger.error("Failed to start CorePlugin: " + name, e);
+            logger.error("Failed to start " + name, e);
         }
     }
 }
 
-function loadPluginBundle(zip: ZipFile) {
-    zip.openEntry("index.js.bundle");
-    const pluginBuffer = zip.readEntry("binary");
-    zip.closeEntry();
+async function loadPlugin(pluginZip: string): Promise<Plugin | null> {
+    logger.info(`Loading plugin from ${pluginZip}.zip`);
+
+    let pluginName: string | null = null;
+
+    try {
+        const zip = new ZipFile(PLUGINS_DIRECTORY + pluginZip, 0, "r");
+
+        zip.openEntry("manifest.json");
+        const manifest = JSON.parse(zip.readEntry("text"));
+        pluginName = manifest.name;
+        zip.closeEntry();
+
+        zip.openEntry("index.js.bundle");
+        const pluginBuffer = zip.readEntry("binary");
+        zip.closeEntry();
+        zip.close();
+
+        const pluginClass = AliuHermes.run(pluginZip, pluginBuffer) as typeof Plugin;
+        if (!(pluginClass?.prototype instanceof Plugin))
+            throw new Error(`Plugin ${pluginName} does not export a valid Plugin`);
+        if (pluginName !== pluginClass.name)
+            throw new Error(`Plugin ${pluginName} must export a class named ${pluginName}`);
+        if (plugins[pluginName])
```

Comment:
Wouldn't this explode when loading a plugin again since it stays in the array, just gets disabled and .enabled set to false?